ballista/rust/executor/src/main.rs

```diff
@@ -154,57 +162,156 @@ async fn main() -> Result<()> {
+            let _shuffle_cleaner_complete = shuffle_cleaner_complete;
+            // As long as the shutdown notification has not been received
+            while !shuffle_cleaner_shutdown.is_shutdown() {
+                tokio::select! {
+                    _ = interval_time.tick() => {
+                            if let Err(e) = clean_shuffle_data_loop(&work_dir, cleanup_ttl as i64).await
+                        {
+                            error!("Ballista executor fail to clean_shuffle_data {:?}", e)
+                        }
+                        },
+                    _ = shuffle_cleaner_shutdown.recv() => {
+                        if let Err(e) = clean_all_shuffle_data(&work_dir).await
+                        {
+                            error!("Ballista executor fail to clean_shuffle_data {:?}", e)
+                        } else {
+                            info!("Shuffle data cleaned.");
+                        }
+                        return;
+                    }
+                };
             }
         });
     }
 
+    let mut service_handlers: FuturesUnordered<JoinHandle<Result<(), BallistaError>>> =
+        FuturesUnordered::new();
+
+    // Channels used to receive stop requests from Executor grpc service.
+    let (stop_send, mut stop_recv) = mpsc::channel::<bool>(10);
+
     match scheduler_policy {
         TaskSchedulingPolicy::PushStaged => {
-            tokio::spawn(executor_server::startup(
-                scheduler,
-                executor.clone(),
-                default_codec,
-            ));
+            service_handlers.push(
+                //If there is executor registration error during startup, return the error and stop early.
+                executor_server::startup(
+                    scheduler,
+                    executor.clone(),
+                    default_codec,
+                    stop_send,
+                    &graceful,
+                )
+                .await?,
+            );
         }
         _ => {
-            tokio::spawn(execution_loop::poll_loop(
+            service_handlers.push(tokio::spawn(execution_loop::poll_loop(
                 scheduler,
                 executor.clone(),
                 default_codec,
-            ));
+            )));
         }
+    };
+    service_handlers.push(tokio::spawn(flight_server_run(
+        addr,
+        graceful.subscribe_for_shutdown(),
+    )));
+
+    // Concurrently run the service checking and listen for the `shutdown` signal and wait for the stop request coming.
+    // The check_services runs until an error is encountered, so under normal circumstances, this `select!` statement runs
+    // until the `shutdown` signal is received or a stop request is coming.
+    tokio::select! {
+        service_val = check_services(&mut service_handlers) => {
+             info!("services stopped with reason {:?}", service_val);
+        },
+        _ = signal::ctrl_c() => {
```

Comment:
My main concern is that `crtl_c` gets called when the process receives a SIGINT on linux. This is how kubernetes/docker will ask a container politely to shutdown. From the docs it does appear SIGINT will trigger `ctrl_c` but I just thought I'd mention something here to see if you tried that @mingmwang .

Reply:
I did not get chance to test it on kubernetes/docker containers yet. I see that some other systems have the similar logic. Let me do some test and will get back on this.

Reply:
https://hyper.rs/guides/server/graceful-shutdown/